File: C#/AIN/DualAINLoop/DualAINLoop.cs
//-----------------------------------------------------------------------------
// DualAINLoop.cs
//
// Demonstrates reading 2 analog inputs (AINs) in a loop from a LabJack.
//
// support@labjack.com
//-----------------------------------------------------------------------------

using System;
using System.Threading;
using LabJack;

namespace DualAINLoop
{
    class DualAINLoop
    {
        static void Main(string[] args)
        {
            DualAINLoop dal = new DualAINLoop();
            dal.performActions();
        }

        public void showErrorMessage(LJM.LJMException e)
        {
            Console.Out.WriteLine("LJMException: " + e.ToString());
            Console.Out.WriteLine(e.StackTrace);
        }

        public void performActions()
        {
            int handle = 0;
            int devType = 0;
            int conType = 0;
            int serNum = 0;
            int ipAddr = 0;
            int port = 0;
            int maxBytesPerMB = 0;
            string ipAddrStr = "";

            try
            {
                //Open first found LabJack
                LJM.OpenS("ANY", "ANY", "ANY", ref handle);  // Any device, Any connection, Any identifier
                //LJM.OpenS("T7", "ANY", "ANY", ref handle);  // T7 device, Any connection, Any identifier
                //LJM.OpenS("T4", "ANY", "ANY", ref handle);  // T4 device, Any connection, Any identifier
                //LJM.Open(LJM.CONSTANTS.dtANY, LJM.CONSTANTS.ctANY, "ANY", ref handle);  // Any device, Any connection, Any identifier

                LJM.GetHandleInfo(handle, ref devType, ref conType, ref serNum, ref ipAddr, ref port, ref maxBytesPerMB);
                LJM.NumberToIP(ipAddr, ref ipAddrStr);
                Console.WriteLine("Opened a LabJack with Device type: " + devType + ", Connection type: " + conType + ",");
                Console.WriteLine("Serial number: " + serNum + ", IP address: " + ipAddrStr + ", Port: " + port + ",");
                Console.WriteLine("Max bytes per MB: " + maxBytesPerMB);

                //Setup and call eWriteNames to configure AINs.
                //AIN0 and AIN1:
                //    Negative Channel = 199 (Single-ended)
                //    Range = +/-10 V
                //        T4 note: Only AIN0-AIN3 can support +/-10 V range.
                //    Resolution index = 0 (default)
                //        T7 note: 0 (default) is index 8, or 9 for Pro.
                //    Settling = 0 (auto)
                int numFrames = 8;
                string[] names = new string[8] { "AIN0_NEGATIVE_CH", "AIN0_RANGE", "AIN0_RESOLUTION_INDEX",
                    "AIN0_SETTLING_US", "AIN1_NEGATIVE_CH", "AIN1_RANGE", "AIN1_RESOLUTION_INDEX",
                    "AIN1_SETTLING_US" };
                double[] aValues = new double[8] {199, 10, 0, 0, 199, 10, 0, 0};
                int errorAddress = 0;
                LJM.eWriteNames(handle, numFrames, names, aValues, ref errorAddress);

                Console.WriteLine("\nSet configuration:");
                for(int i = 0; i < numFrames; i++)
                {
                    Console.WriteLine("  " + names[i] + " : " + aValues[i]);
                }

                //Setup and call eReadNames to read AINs.
                numFrames = 2;
                names = new string[2] {"AIN0", "AIN1"};
                aValues = new double[2] {0, 0};
                
                Console.WriteLine("\nStarting read loop.  Press a key to stop.");
                while(!Console.KeyAvailable)
                {
                    LJM.eReadNames(handle, numFrames, names, aValues, ref errorAddress);
                    Console.WriteLine("\n" + names[0] + " : " + aValues[0].ToString("F4") + " V, " + names[1] + " : " + aValues[1].ToString("F4") + " V");
                    Thread.Sleep(1000); //Wait 1 second
                }
            }
            catch (LJM.LJMException e)
            {
                showErrorMessage(e);
            }

            LJM.CloseAll();  //Close all handles

            Console.WriteLine("\nDone.\nPress the enter key to exit.");
            Console.ReadLine();  // Pause for user
        }
    }
}

File: VB/AIN/DualAINLoop/DualAINLoop.vb
'------------------------------------------------------------------------------
' DualAINLoop.vb
'
' Demonstrates reading 2 analog inputs (AINs) in a loop.
'
' support@labjack.com
'------------------------------------------------------------------------------
Option Explicit On

Imports LabJack
Imports System.Threading

Module DualAINLoop

    Sub showErrorMessage(ByVal e As LJM.LJMException)
        Console.WriteLine("LJMException: " & e.ToString)
        Console.WriteLine(e.StackTrace)
    End Sub

    Sub displayHandleInfo(ByVal handle As Integer)
        Dim devType As Integer
        Dim conType As Integer
        Dim serNum As Integer
        Dim ipAddr As Integer
        Dim port As Integer
        Dim maxBytesPerMB As Integer
        Dim ipAddrStr As String = ""

        LJM.GetHandleInfo(handle, devType, conType, serNum, ipAddr, port, _
                          maxBytesPerMB)
        LJM.NumberToIP(ipAddr, ipAddrStr)
        Console.WriteLine("Opened a LabJack with Device type: " & devType & _
                          ", Connection type: " & conType & ",")
        Console.WriteLine("Serial number: " & serNum & ", IP address: " & _
                          ipAddrStr & ", Port: " & port & ",")
        Console.WriteLine("Max bytes per MB: " & maxBytesPerMB)
    End Sub

    Sub Main()
        Dim handle As Integer
        Dim numFrames As Integer
        Dim aNames() As String
        Dim aValues() As Double
        Dim errAddr As Integer = -1

        Try
            ' Open first found LabJack
            LJM.OpenS("ANY", "ANY", "ANY", handle)
            'LJM.Open(LJM.CONSTANTS.dtANY, LJM.CONSTANTS.ctANY, "ANY", handle)

            displayHandleInfo(handle)

            ' Setup and call eWriteNames to configure AINs.
            ' Setting AIN0-1 Negative Channel to 199 (Single-ended), Range to
            ' +-10 V, Resolution index to 0 (default: index 8 or 9 for Pro) and
            ' Settling to 0 (automatic).
            numFrames = 6
            ReDim aNames(numFrames - 1)
            aNames(0) = "AIN0_NEGATIVE_CH"
            aNames(1) = "AIN0_RANGE"
            aNames(2) = "AIN0_RESOLUTION_INDEX"
            aNames(3) = "AIN1_NEGATIVE_CH"
            aNames(4) = "AIN1_RANGE"
            aNames(5) = "AIN1_RESOLUTION_INDEX"
            ReDim aValues(numFrames - 1)
            aValues(0) = 199
            aValues(1) = 10
            aValues(2) = 0
            aValues(3) = 199
            aValues(4) = 10
            aValues(5) = 0
            LJM.eWriteNames(handle, numFrames, aNames, aValues, errAddr)

            Console.WriteLine("")
            Console.WriteLine("Set configuration:")
            For i = 0 To numFrames - 1
                Console.WriteLine("  " & aNames(i) & " : " & aValues(i))
            Next

            ' Setup and call eReadNames to read AINs.
            numFrames = 2
            ReDim aNames(numFrames - 1)
            aNames(0) = "AIN0"
            aNames(1) = "AIN1"
            ReDim aValues(numFrames - 1)
            aValues(0) = 0
            aValues(1) = 0
            
            Console.WriteLine("")
            Console.WriteLine("Starting read loop.  Press a key to stop.")
            While Console.KeyAvailable = False
                LJM.eReadNames(handle, numFrames, aNames, aValues, errAddr)
                Console.WriteLine("")
                Console.WriteLine(aNames(0) & " : " & _
                                  aValues(0).ToString("F4") & " V, " & _
                                  aNames(1) & " : " & _
                                  aValues(1).ToString("F4") & " V")
                Thread.Sleep(1000) ' Wait 1 second
            End While
        Catch ljme As LJM.LJMException
            showErrorMessage(ljme)
        End Try

        LJM.CloseAll() ' Close all handles

        Console.WriteLine("")
        Console.WriteLine("Done.")
        Console.WriteLine("Press the enter key to exit.")
        Console.ReadLine() ' Pause for user
    End Sub


End Module
